# Ejercicio 3: Redes Bayesianas

Factorizaci&oacute;n de la distribuci&oacute;n conjunta basada en la estructura de la red bayesiana:

$$P(G, M, B, C) = P(G) \cdot P(M \mid G) \cdot P(B \mid M) \cdot P(C \mid M)$$

In [1]:
from fractions import Fraction
import pandas as pd

# Probabilidades dadas en el enunciado
P_G = {"+": Fraction(1, 10), "-": Fraction(9, 10)}

P_M_given_G = {
    "+": {"+": Fraction(2, 3), "-": Fraction(1, 3)},
    "-": {"+": Fraction(1, 4), "-": Fraction(3, 4)},
}

P_B_given_M = {
    "+": {"+": Fraction(2, 5), "-": Fraction(3, 5)},
    "-": {"+": Fraction(1, 5), "-": Fraction(4, 5)},
}

P_C_given_M = {
    "+": {"+": Fraction(1, 4), "-": Fraction(3, 4)},
    "-": {"+": Fraction(1, 2), "-": Fraction(1, 2)},
}

# Cálculo de probabilidad conjunta
def joint(g, m, b, c):
    return P_G[g] * P_M_given_G[g][m] * P_B_given_M[m][b] * P_C_given_M[m][c]

# Generar distribución completa
rows = []
for g in ["+", "-"]:
    for m in ["+", "-"]:
        for b in ["+", "-"]:
            for c in ["+", "-"]:
                prob_frac = joint(g, m, b, c)
                prob_dec = float(prob_frac)
                rows.append({
                    "G (Presid.)": g,
                    "M (Legaliz.)": m,
                    "B (Presup.)": b,
                    "C (Clases)": c,
                    "Probabilidad (Fracción)": str(prob_frac),
                    "Probabilidad (Decimal)": f"{prob_dec:.6f}"
                })

df = pd.DataFrame(rows)
total_prob = sum(joint(r["G (Presid.)"], r["M (Legaliz.)"], r["B (Presup.)"], r["C (Clases)"]) for r in rows)
print(f"Suma total de la conjunta: {total_prob} (Consistencia verificada)\n")
df

Suma total de la conjunta: 1 (Consistencia verificada)



,G (Presid.),M (Legaliz.),B (Presup.),C (Clases),Probabilidad (Fracción),Probabilidad (Decimal)
0,+,+,+,+,1/150,0.006667
1,+,+,+,-,1/50,0.020000
2,+,+,-,+,1/100,0.010000
3,+,+,-,-,3/100,0.030000
4,+,-,+,+,1/300,0.003333
5,+,-,+,-,1/300,0.003333
6,+,-,-,+,1/75,0.013333
7,+,-,-,-,1/75,0.013333
8,-,+,+,+,9/400,0.022500
9,-,+,+,-,27/400,0.067500


## Resultados: Valores Faltantes en la Tabla

| Probabilidad Solicitada | Expresi&oacute;n Fraccional | Valor Decimal |
| :--- | :---: | :---: |
| **$P(+g, +m, +b, -c)$** | **$1/50$** | **$0.020000$** |
| **$P(+g, +m, -b, -c)$** | **$3/100$** | **$0.030000$** |
| **$P(+g, -m, -b, +c)$** | **$1/75$** | **$0.013333$** |
| **$P(-g, +m, +b, +c)$** | **$9/400$** | **$0.022500$** |
| **$P(-g, +m, -b, +c)$** | **$27/800$** | **$0.033750$** |
| **$P(-g, -m, -b, +c)$** | **$27/100$** | **$0.270000$** |